### PR TITLE
Minimum size not compatible with template package

### PR DESCRIPTION
**Pull request by alexandreStein**
- go 1.8
- Linux 64
- I using the package with templates
- I expect to see the template as it should :smiley:
- But I get the template with some special chars. It looks like bad decoding

The problem comes from the size limitation.

If the template package is used, it can calls Write multiple times with variable length.
In any case the response writer must be the same for the all response inside the same request. So the package as is, is not good.

I made some modifications to prevent using different response writers on successive calls.

But the size limitation become pretty pointless with template or other handlers calling multiple times the Write function.

### CODE REVIEW

File: gzip.go [hunk 1]
```diff
@@ -65,17 +72,29 @@ func addLevelPool(level int) {
 // writers, so don't forget to do that.
```
Comment (tonglil) [inline gzip.go]: Please update the func doc with the minSize conditioning.

File: gzip.go [hunk 2]
```diff
@@ -65,17 +72,29 @@ func addLevelPool(level int) {
...
-	index int // Index for gzipWriterPools.
-	gw    *gzip.Writer
+	index        int // Index for gzipWriterPools.
+	gw           *gzip.Writer
+	minSize      int
+	chosenWriter io.Writer // after the first Write call the ResponseWriter must use the same writer
 }
 
 // Write appends data to the gzip writer.
 func (w *GzipResponseWriter) Write(b []byte) (int, error) {
```
Comment (tonglil) [inline gzip.go]: Same here.

File: gzip.go [hunk 3]
```diff
@@ -162,9 +181,18 @@ func MustNewGzipLevelHandler(level int) func(http.Handler) http.Handler {
 // if an invalid gzip compression level is given, so if one can ensure the level
 // is valid, the returned error can be safely ignored.
 func NewGzipLevelHandler(level int) (func(http.Handler) http.Handler, error) {
```
Comment (tonglil) [inline gzip.go]: Same here.

File: gzip.go [hunk 4]
```diff
@@ -162,9 +181,18 @@ func MustNewGzipLevelHandler(level int) func(http.Handler) http.Handler {
...
+	return NewGzipLevelAndMinSize(level, DefaultMinSize)
+}
+
+// NewGzipLevelAndMinSize do as above but let caller specify the minimum size
```
Comment (tonglil) [inline gzip.go]: Can you specify "NewGzipLevelHandler" instead of "do as above"?

Also please end the doc comment with a period.

File: gzip.go [hunk 5]
```diff
@@ -63,19 +70,34 @@ func addLevelPool(level int) {
...
 
 // Write appends data to the gzip writer.
 func (w *GzipResponseWriter) Write(b []byte) (int, error) {
+	// If a partial response have been send, the next response will use the same writer.
+	if w.chosenWriter != nil {
+		return w.chosenWriter.Write(b)
+	}
+
+	// If b is smaller than the minSize, the regular writer is used.
+	if len(b) < w.minSize {
```
Comment (adammck) [inline gzip.go]: This isn't sufficient. We must keep track of the total number of bytes which have been written, because `Write` can be called many times. We can't decide whether to gzip the response or not until the headers are written; and even then, we can only make the call based on how many bytes have been written _so far_.

Comment (alexandreStein) [inline gzip.go]: I did it again.
I hop it is all good this time.

First time, I forgot about Write which can be call multiple times.
And I just patched my crappy implementation.

This one looks better to me.

**Comment by jprobinson** [comment]
Looks like another recent PR may have caused some merge conflicts. Please clean those up and, if @adammck also gives blessing, I'll pull this in.

**Comment by jprobinson** [comment]
Yo, @adammck! Have a sec to put your :eyes: on this one? I'd like your 👌 before merging.

File: gzip.go [hunk 6]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
 // writers, so don't forget to do that.
+// It can be configured to skip response smaller than minSize.
 type GzipResponseWriter struct {
 	http.ResponseWriter
 	index int // Index for gzipWriterPools.
 	gw    *gzip.Writer
+
+	code int // Saves the WriteHeader value.
+
+	minSize      int    // Specifed the minimum response size to gzip. Only the first Write call is checked.
```
Comment (adammck) [inline gzip.go]: Just to clarify, this is the minimum _uncompressed_ bytes, right?

File: .travis.yml [hunk 7]
```diff
@@ -1,5 +1,5 @@
 language: go
 
 go:
-  - 1.7
+  - 1.x
```
Comment (adammck) [inline .travis.yml]: This change should really be a separate PR...

File: gzip.go [hunk 8]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
 
 // Write appends data to the gzip writer.
 func (w *GzipResponseWriter) Write(b []byte) (int, error) {
-	// Lazily create the gzip.Writer, this allows empty bodies to be actually
-	// empty, for example in the case of status code 204 (no content).
-	if w.gw == nil {
-		w.init()
-	}
+	// Update the numbers of bytes written.
+	w.bytesWritten += len(b)
```
Comment (adammck) [inline gzip.go]: This should occur immediately after we've written the bytes.

File: gzip.go [hunk 9]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
 func (w *GzipResponseWriter) WriteHeader(code int) {
-	if w.gw == nil &&
-		code != http.StatusNotModified && code != http.StatusNoContent {
-		w.init()
+	w.code = code
+}
+
+// writeHeader uses the saved code to send it to the ResponseWriter.
+func (w *GzipResponseWriter) writeHeader() {
+	// debug.PrintStack()
```
Comment (adammck) [inline gzip.go]: Please remove this.

File: gzip.go [hunk 10]
```diff
@@ -102,15 +166,17 @@ func (w *GzipResponseWriter) init() {
...
-	w.ResponseWriter.Header().Del(contentLength)
 }
 
 // Close will close the gzip.Writer and will put it back in the gzipWriterPool.
 func (w *GzipResponseWriter) Close() error {
+	// Buffer not nil means the regular response it returned.
+	if w.buff != nil {
+		w.writeHeader()
+		w.ResponseWriter.Write(w.buff)
+		return nil
```
Comment (adammck) [inline gzip.go]: Minor point: We don't need to return early here, because we'll do so immediately below, since w.gw is nil. I think it's better that way, since we don't have to rely on the (w.buff == nil) == (w.gw == nil) behavior never changing.

File: gzip.go [hunk 11]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
+
+		// Add the buffered length to the new write length.
+		n += n1
+		return n, err
+	}
+
+	// When arrived here, it's to save the writes into the buffer until it
+	// goes over the minSize or it closes.
+	w.buff = append(w.buff, b...)
+	return 0, nil
```
Comment (adammck) [inline gzip.go]: Is it correct to say that zero bytes were written, here? They were written into the buffer, after all. The fact that they didn't reach the wire yet is surely not of concern to the caller.

File: gzip.go [hunk 12]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
 		// If content type is not set, infer it from the uncompressed body.
 		w.Header().Set(contentType, http.DetectContentType(b))
 	}
-	return w.gw.Write(b)
+
+	// If the global write is bigger than the minSize, compression is used.
+	// Otherwise it save the write into a buffer and send it with the regular
+	// ResponseWriter at close.
+	if w.bytesWritten > w.minSize {
+		w.Header().Set(contentEncoding, "gzip")
```
Comment (adammck) [inline gzip.go]: There's really no need to set this header on every Write after bytesWritten exceeds minSize. It only needs to happen the first time. We can add a `isGzipped` (or whatever) field to the writer to keep track.

File: gzip.go [hunk 13]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
+
+			// Saved the buffed length.
+			n = n1
+		}
+
+		w.buff = nil
+
+		if w.gw == nil {
+			w.init()
+		}
```
Comment (adammck) [inline gzip.go]: We do this on :112, too. Is there some reason that we can't do it after writeHeader, since we already know that the response is going to be gzipped?

File: gzip.go [hunk 14]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
+
+		n1, err := w.gw.Write(b)
+		if err != nil {
+			return n, err
```
Comment (adammck) [inline gzip.go]: Should be `n+n1`, since some bytes can be written even if the Write failed.

File: gzip.go [hunk 15]
```diff
@@ -163,9 +229,18 @@ func MustNewGzipLevelHandler(level int) func(http.Handler) http.Handler {
 // if an invalid gzip compression level is given, so if one can ensure the level
 // is valid, the returned error can be safely ignored.
 func NewGzipLevelHandler(level int) (func(http.Handler) http.Handler, error) {
+	return NewGzipLevelAndMinSize(level, DefaultMinSize)
+}
+
+// NewGzipLevelAndMinSize behave as NewGzipLevelHandler except it let the caller
+// specify the minimum size before compression.
+func NewGzipLevelAndMinSize(level, askedMinSize int) (func(http.Handler) http.Handler, error) {
```
Comment (adammck) [inline gzip.go]: askedMinSize can just be minSize.

File: gzip_test.go [hunk 16]
```diff
@@ -134,8 +134,8 @@ func TestGzipHandlerNoBody(t *testing.T) {
 		// Body must be empty.
 		{http.StatusNoContent, "", 0},
 		{http.StatusNotModified, "", 0},
-		// Body is going to get gzip'd no matter what.
-		{http.StatusOK, "gzip", 23},
+		// // Body is going to get gzip'd no matter what.
+		// {http.StatusOK, "gzip", 23},
```
Comment (adammck) [inline gzip_test.go]: Please remove commented code.

File: gzip.go [hunk 17]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
+// It can be configured to skip response smaller than minSize.
 type GzipResponseWriter struct {
 	http.ResponseWriter
 	index int // Index for gzipWriterPools.
 	gw    *gzip.Writer
+
+	code int // Saves the WriteHeader value.
+
+	minSize      int    // Specifed the minimum response size to gzip. Only the first Write call is checked.
+	buff         []byte // Holds the first part of the write before the reaching the minSize or the end of the write.
```
Comment (adammck) [inline gzip.go]: These are more commonly named simply `buf`.

File: gzip.go [hunk 18]
```diff
@@ -63,35 +69,93 @@ func addLevelPool(level int) {
...
+	bytesWritten int    // Keep trace of the numbers of bytes written.
 }
 
 // Write appends data to the gzip writer.
 func (w *GzipResponseWriter) Write(b []byte) (int, error) {
```
Comment (adammck) [inline gzip.go]: I think we need to reconsider this function. It's quite busy and stateful. It would be better to split it into two parts:

This func is:
1. If gw != nil (i.e. we are gzipping), write and return
2. Append to buf
3. If len(buf) has exceeded minGzipBytes, call startGzip

Where `startGzip` is:
1. Write headers
2. Set gw
3. Flush buf

Am I missing anything here?

**Comment by adammck** [comment]
Sorry for the last-minute review, @jprobinson, but you did ask :smile:
Feel free to dismiss if you'd prefer to get this merged and tidy things up later.

File: gzip.go (hunk 11 above)
Comment (alexandreStein) [inline gzip.go]: If the response is not compressed, the caller get a response with **0** and **nil** which could looks like an error but which is not.
Because the actual write happens at close we can't return to caller the numbers of bytes written before GZIP.

I'm not sure what is the best option. 
Discussion is open. :smiley:

File: gzip.go (hunk 18 above)
Comment (alexandreStein) [inline gzip.go]: Your are is right.
This function is now splitted into two cleaner and shorter functions.

Does it look better to you?

Comment (adammck) [inline gzip.go]: Much better, thank you!

**Comment by adammck** [comment]
A bit heavy on the comments, but otherwise LGTM :+1:
Thanks for the quick turnaround on those changes!

File: gzip.go [hunk 19]
```diff
@@ -63,35 +69,84 @@ func addLevelPool(level int) {
...
+	w.code = code
+}
+
+// writeHeader uses the saved code to send it to the ResponseWriter.
+func (w *GzipResponseWriter) writeHeader() {
+	if w.code == 0 {
+		w.code = http.StatusOK
 	}
-	w.ResponseWriter.WriteHeader(code)
+	w.WriteHeader(w.code)
```
Comment (adammck) [inline gzip.go]: Oops, missed this. Fixed by #38.
